app/controllers/course/stage/instructions.ts

```diff
@@ -47,16 +48,32 @@ export default class CourseStageInstructionsController extends Controller {
     return this.model.courseStage.prerequisiteInstructionsMarkdownFor(this.model.activeRepository);
   }
 
+  get shouldHideTestRunnerCardBeforeUserHasSubmitted() {
+    return this.featureFlags.hideTestRunnerCardBeforeUserHasSubmitted;
+  }
+
   get shouldShowFeedbackPrompt() {
     return !this.currentStep.courseStage.isFirst && this.currentStep.status === 'complete';
   }
 
+  get shouldShowLinkToForum() {
+    return (
+      this.isCurrentStage &&
+      this.currentStep.status !== 'complete' &&
+      this.currentStep.testsStatus !== 'passed' &&
+      this.shouldHideTestRunnerCardBeforeUserHasSubmitted
+    );
+  }
+
   get shouldShowPrerequisites() {
     return !!this.prerequisiteInstructionsMarkdown;
   }
 
   get shouldShowTestRunnerCard() {
-    return this.isCurrentStage && this.currentStep.status !== 'complete';
+    const shouldHideTestRunnerCard =
+      this.model.courseStage.isFirst && this.model.activeRepository.submissions.length <= 1 && this.shouldHideTestRunnerCardBeforeUserHasSubmitted;
+
+    return this.isCurrentStage && this.currentStep.status !== 'complete' && !shouldHideTestRunnerCard;
```

Comment:
Lots of conditions here, would recommend simplifying by splitting out the various cases: 

```
get shouldShowTestRunnerCard() {
  if (!this.isCurrentStage) {
    return false
  }
  
  if (this.currentStep.status === 'complete') {
    return false;
  }
  
  if (this.model.courseStage.isFirst) {
    // For stage 1, we hide the test runner card until the user's submission.
    return ...
  } else {
    // For other stages, we always show the test runner card
    return true;
  }
}
```